### ISSUE TITLE
116 nodes gradually marked unhealthy over the course of 15 minutes

### ISSUE THREAD

**Issue by bjornswift** [kubernetes/kubernetes#46641]
**Kubernetes version**:
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:33:11Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2+coreos.0", GitCommit:"79fee581ce4a35b7791fdd92e0fc97e02ef1d5c0", GitTreeState:"clean", BuildDate:"2017-04-19T23:13:34Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```
**Environment**:
- **Cloud provider or hardware configuration**: aws
- **OS**:
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1353.7.0
VERSION_ID=1353.7.0
BUILD_ID=2017-04-26-2154

- **Kernel**: Linux 4.9.24-coreos #1 SMP Wed Apr 26 21:44:23 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux

- **Install tools**: kube-aws 0.9.6

------------

We're investigating an incident we had last week where 116 nodes in a 119 node cluster moved to NodeNotReady. Nodes moved into NodeNotReady one by one, over a period of 15 minutes. Then, all nodes became healthy at around the same time.

Here's what we know:
 1) We're running Kubernetes v1.6.2, spun up with kube-aws v0.9.6 on CoreOS's stable release branch.

 2) While the nodes were unhealthy, their pods remained running -- but stopped receiving incoming traffic. The service fronting the pods returned an empty result set (no matching pods). The application running inside the pod did not show signs of network trouble, appeared able to service already-established connections.

 3) Kubernetes initialized a pod removal for nodes that had been unhealthy for 5 minutes (the pod eviction time).

 4) The pod removal was processed only when the nodes became healthy again (presumably because the controller could not communicate with the kubelet).

 5) The closest we've gotten to root cause is that it's "something related to iptables". The nodes' dmesg contains:

```
ip-10-0-7-5.ec2.internal.dmesg:[Fri May 26 00:05:34 2017] Netfilter messages via NETLINK v0.30.
ip-10-0-7-5.ec2.internal.dmesg-[Fri May 26 00:05:34 2017] ctnetlink v0.93: registering with nfnetlink.
```

And then a subset of the nodes failed ensuring correct rules were running:

```
May 26 00:06:37 ip-10-0-7-5.ec2.internal kubelet-wrapper[1911]: E0526 00:06:37.336142    1911 kubelet_network.go:412] Failed to ensure marking rule for KUBE-MARK-MASQ: error checking rule: exit status 4: iptables: Resource temporarily unavailable.
May 26 00:20:38 ip-10-0-7-5.ec2.internal kubelet-wrapper[1911]: E0526 00:20:38.629536    1911 kubelet_network.go:378] Failed to ensure marking rule for KUBE-MARK-DROP: error checking rule: exit status 4: iptables: Resource temporarily unavailable.
```

  6) All nodes printed the Netfilter message between 2017-05-26T00:05:34Z - 2017-05-26T00:10:02Z, and all nodes were marked unhealthy between 2017-05-26T00:05:41Z - 2017-05-26T00:20:06Z. There is not a clear correlation between which nodes printed the netfilter message first, and which were detected unhealthy first.

We don't know where to go from here, and would appreciate pointers. We've collected the journal and dmesg on all nodes, the kubectl event stream, as well as the api-server, controller and scheduler logs. Don't believe they contain any sensitive data, but would prefer to share over a private channel, if possible.

Are there any other logs we should be gathering before tearing down the cluster?

**Comment by snarayanank2** [comment]
We just ran into a similar problem - this has happened twice in the last 2 days. We are on aws, use kube-aws 0.9.4. AFAICT, the sequence of events is:

* Netfilter messages appear
* Node gets marked unhealthy
* We see connection timed out messages like the following:
```
Jun 01 19:36:14 ip-10-0-0-114.ap-southeast-1.compute.internal kernel: Netfilter messages via NETLINK v0.30.
Jun 01 19:36:14 ip-10-0-0-114.ap-southeast-1.compute.internal kernel: ctnetlink v0.93: registering with nfnetlink.
```
* Also see messages like this:
```
Jun 01 19:51:35 ip-10-0-0-114.ap-southeast-1.compute.internal systemd-networkd[780]: veth35277ad0: Lost carrier
Jun 01 19:51:57 ip-10-0-0-114.ap-southeast-1.compute.internal kernel: IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
Jun 01 19:51:57 ip-10-0-0-114.ap-southeast-1.compute.internal systemd-timesyncd[675]: Network configuration changed, trying to establish connection.
Jun 01 19:51:57 ip-10-0-0-114.ap-southeast-1.compute.internal systemd-timesyncd[675]: Synchronized to time server 128.199.134.40:123 (1.coreos.pool.ntp.org).
Jun 01 19:51:57 ip-10-0-0-114.ap-southeast-1.compute.internal kernel: IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
Jun 01 19:51:57 ip-10-0-0-114.ap-southeast-1.compute.internal systemd-networkd[780]: veth288489af: Gained carrier
```
* After about 25 mins everything is back to normal

Any help would be greatly appreciated!

**Comment by bjornswift** [comment]
We just had another incident where 60 nodes (in a 125 node cluster) were marked NotReady, before recovering ~15 minutes later.

We're going to temporarily increase the pod eviction time to mitigate, but this is currently our biggest stability risk.

Any pointers as to what we could investigate -- or what we could provide for further clarification -- would be greatly appreciated.

**Comment by kargakis** [comment]
@kubernetes/sig-scalability-misc

**Comment by bjornswift** [comment]
We had another event, about 45 minutes later, this time lasting slightly longer -- long enough for ~half of our pods to be terminated (evicted).

The remaining pods were unable to handle the increased traffic and we suffered downtime as we brought back capacity and got things to stabilize.

**Comment by timothysc** [comment]
FWIW I have not seen this behavior on a number of other stacks.  /cc @philips for routing.

**Comment by rodawg** [comment]
I believe I'm experiencing the same problem -- we've gone through the same exact number of mini outages/events today.

However, I'm not seeing the netfilter errors, only errors matching `Failed to ensure marking rule for KUBE-MARK-MASQ`, etc. 

I'm running CentOS 7 on AWS with the kubernetes version:
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:36:33Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.7", GitCommit:"8eb75a5810cba92ccad845ca360cf924f2385881", GitTreeState:"clean", BuildDate:"2017-04-27T09:42:05Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by bjornswift** [comment]
We had another event today. Similar to yesterday's event. Nodes were NotReady for ~20 minutes, then recovered. Around 20 minutes later they were again marked NotReady for ~20 minutes.

We had increased the pod eviction time, so we didn't loose pods. However, as all DNS pods were on nodes that were NotReady, Kubernetes updated the dns service to have no endpoints, and we couldn't resolve DNS during the first incident.

**Comment by bjornswift** [comment]
We had another event this morning, third day in a row.

 - Again, in two groups.
 - The two groups cover the whole cluster (all nodes except the controller).
 - This time the events were spaced further apart, or around 2 hours (let's call these `t1` and `t2`).
 - All the `Netfilter messages via NETLINK` messages were printed at `t1`, also for the nodes that went into NotReady state at `t2`.
 - Some nodes are reporting `nf_conntrack: table full, dropping packet`. There in an increase in these around the events, but we haven't seen a really strong correlation yet. But, we'll try increasing the max and start graphing the usage.

One more data point: We decided to spin up a backup cluster and have that ready in case we had issues today. This backup cluster was started ~3.5 hour after the primary cluster -- and it just had it's first wave of NotReady nodes, ~3.5 hours after the primary cluster. So, today, both clusters (with wildly different network traffic, but the same combination of pods) hit this ~14 hours after being started.

**Comment by bjornswift** [comment]
We've been having these events daily now since last Wednesday (6 days).

@snarayanank2 and @rodawg, are you also still experiencing the issue?

Some notes:
 - Have ruled out `conntrack_max`. A silent test cluster also experienced the kube outage, even though `conntrack` was low.
 - Have spun up three test clusters running older versions of our code (older docker images) to rule out changes in our software. All three test clusters experienced the outage.
 - Interestingly, all three test clusters experienced the issue after roughly the same amount of time. This could indicate that something builds up and then falls over. Not seen anything in logs yet, but having trouble getting to the etcd logs.
 - Have not yet tried older AMIs of CoreOs. CoreOs updated their stable branch on May 31st, which is a week after our first event, and a week before we started having daily issues - so does not look correlated.

**Comment by rodawg** [comment]
@bjornswift We've seen it occur over the last few days as well, though it hasn't happened today (yet).

I in the middle of individually testing a few changes to see what prevents the issue from occurring.

* Updating to the latest available CentOS 7 AMI on the AWS Marketplace
* I've enabled kube-apiserver audit logging to see if there's any pattern to the API requests sent to the cluster
* Removed a few unnecessary kubernetes services, which reduced the number of iptables rules from about 8000 to 3000
* Increased the max ulimit of the cluster components to match up with the [getting started](https://kubernetes.io/docs/getting-started-guides/scratch/#docker) docs
* Updating our Docker version to 1.12.6

We're running 2 clusters, a large 170 node one and small 6 node one. I've seen the event occur less often on the smaller cluster, and not for a few days.

**Comment by bjornswift** [comment]
@rodawg Had you been running the same versions of Kubernetes/CentOs for a while before experiencing the issue?

In our case, we'd been running the same version of everything since early May -- but your and my environments are quite different. You're on Kubernetes 1.5.x, we're 1.6.2. You're on CentOs, we're on CoreOs. We're both on AWS, what else is common?

Us:
 - We've been experiencing this daily since June 7th
 - Our instance type is r4.2xlarge and we're running in us-east-1
 - Seems to occur ~14 - ~24 hours into the lifetime of the cluster
 - Cluster size seems to affect how long this takes to trigger.
 - We move from one Kubernetes cluster to another daily.
 - We're running 3 etcd nodes (version 2), set up by CoreOS' kube-aws

Do you see any anomalies on etcd on your end? I'm having trouble getting to our etcd logs, so that's one piece I haven't been able to fully investigate.

**Comment by rodawg** [comment]
* We've also been experiencing this daily since June 7th, but the last occurrence was Monday AM.
* Our instance type is c3.4xlarge on us-east-1
* We're not bringing up a new cluster, only restarting the kubelet process
* The time between events is about the same, though it hasn't occurred since I reduced the overall number of services and iptables rules
* The smaller cluster doesn't experience this nearly as often
* We're running a HA etcd/kube-apiserver setup as well built from scratch and managed by Saltstack

I haven't seen any anomalies on the etcd end.

**Comment by snarayanank2** [comment]
@bjornswift we have not run into the issue after adjusting some parameters. My colleague will add a note on what the specific parameters we changed.

**Comment by kavadithya** [comment]
Here's a solution that worked for us (@@snarayanank2's team). 
1. set node-status-update-frequency to 20 seconds on the controller node
2. node-monitor-grace-period to 300 seconds on your controller and worker nodes
The kubelet service has to be started with the 1st setting, and kube-controller-manager has to be started with the 2nd setting. 
We haven't seen the issue for the past 10 days after putting in the above settings.

**Comment by bjornswift** [comment]
@kavadithya / @snarayanank2 thanks!

I'm going to apply these changes to our clusters and see how that goes. Can see how this would mitigate impact (i.e. the cluster is given longer to recover from events), but wonder whether there's still some underlying issue that causes the network event to occur in the first place.

Will apply and report back, thanks!

**Comment by bjornswift** [comment]
@rodawg:

> We're not bringing up a new cluster, only restarting the kubelet process

Interesting. So had your cluster been around for several days before June 7th, and then this started to occur daily? That would really rule out a change in software configuration.

We also see the eth0 network log messages that @snarayanank2 reported, ~10-20 minutes after the netfilter/iptables message. I've cut a support ticket to aws to see if they see anything of relevance on the host machine.

**Comment by rodawg** [comment]
> Interesting. So had your cluster been around for several days before June 7th, and then this started to occur daily? That would really rule out a change in software configuration.

Nothing had changed for over a week before the first occurrence.

I also see the eth0 network log messages but none of the other messages.

**Comment by rodawg** [comment]
@bjornswift any luck with those settings? I haven't seen another occurrence until an hour ago, without those settings mentioned earlier.

**Comment by bjornswift** [comment]
@rodawg Unfortunately these settings haven't helped us. We've seen incidents on the 16th, 17th, 18th, 22nd, 24th, 25th and then now right around midnight June 26th (4 hours ago).

We've managed to mitigate most impact by extending the pod eviction time. That means that pods don't get deleted (and continue to service established connections, which is important for us). However, they get removed as endpoints from the Kubernetes service, and thus don't get any new incoming traffic.

The next things we're going to try is deploying network probes on the cluster to try and get a better understanding on the networking event.

**Comment by Dao007forever** [comment]
https://github.com/kubernetes/kubernetes/issues/46641

Looks like the same problem.

**Comment by bjornswift** [comment]
Closing in favor of #41916 and #48638
